src/Event.php

```diff
@@ -80,24 +98,30 @@ public function getTarget()
      *
      * Overwrites parameters
      *
-     * @param  array|ArrayAccess|object $params
+     * @param array|ArrayAccess|object $params
+     * @template NewTParams of array|ArrayAccess|object
+     * @psalm-param NewTParams $params
+     * @psalm-this-out self<TTarget, NewTParams>
      * @throws Exception\InvalidArgumentException
      */
     public function setParams($params)
     {
+        /** @psalm-suppress DocblockTypeContradiction Sanity check to actually enforce docblock. */
         if (! is_array($params) && ! is_object($params)) {
             throw new Exception\InvalidArgumentException(
                 sprintf('Event parameters must be an array or object; received "%s"', gettype($params))
             );
         }
 
+        /** @psalm-suppress InvalidPropertyAssignmentValue Pretty sure this is correct after this-out. */
         $this->params = $params;
```

Comment:
I'm just as unsure =S

It looks like it only updates the template type after the function returns. So within the function it will still assume the previous type.

Reply:
Yeah, same experience: this is good as-is.